File: elixir/pig_latin.ex
defmodule PigLatin do
  @doc """
  Given a `phrase`, translate it a word at a time to Pig Latin.
  """
  @spec translate(phrase :: String.t()) :: String.t()
  def translate(phrase) do
    if String.contains?(phrase, " ") do
      phrase
      |> String.split(" ")
      |> Enum.map(&translate/1)
      |> Enum.join(" ")
    else
      cond do
        Regex.match?(~r/^[xy][^aeiuo]/, phrase) ->
          phrase <> "ay"

        Regex.match?(~r/^[aeiou]/, phrase) ->
          phrase <> "ay"

        Regex.match?(~r/^[^aeiou]*qu/, phrase) ->
          [_, a, b] = Regex.run(~r/^([^aeiou]*qu)(.*)/, phrase)
          b <> a <> "ay"

        Regex.match?(~r/^[^aeiou]+[aeiou]/, phrase) ->
          [_, a, b] = Regex.run(~r/^([^aeiou]+)([aeiou].*)/, phrase)
          b <> a <> "ay"

        Regex.match?(~r/^(.*)y/, phrase) ->
          [_, a, b] = Regex.run(~r/^(.*)(y.*)/, phrase)
          b <> a <> "ay"
      end
    end
  end
end

File: go/pig_latin.go
package piglatin

import (
	"regexp"
	"strings"
)

func Word(word string) string {
	re := regexp.MustCompile(`^([aeiuo]|xr|yt)`)
	if re.MatchString(word) {
		return word + "ay"
	}

	re = regexp.MustCompile(`^([^aeiou]*qu)(.+)$`)
	if match := re.FindStringSubmatch(word); match != nil {
		return match[2] + match[1] + "ay"
	}

	re = regexp.MustCompile(`^([^aeiou]+|.)(y.+)$`)
	if match := re.FindStringSubmatch(word); match != nil {
		return match[2] + match[1] + "ay"
	}

	re = regexp.MustCompile(`^([^aeiou]+)(.+)$`)
	if match := re.FindStringSubmatch(word); match != nil {
		return match[2] + match[1] + "ay"
	}

	return word
}

func Sentence(sentence string) string {
	words := strings.Split(sentence, " ")
	arr := make([]string, len(words))
	for i, w := range words {
		arr[i] = Word(w)
	}
	return strings.Join(arr, " ")
}
